File: lib/poker_node/holdem/deal.rb
module PokerNode::Holdem
  class Deal
    attr_reader :flop
    attr_reader :turn_card
    attr_reader :river_card

    def turn
      flop + [turn_card]
    end

    def river
      turn + [river_card]
    end

    attr_reader :hands
    attr_reader :hole_cards
    attr_reader :holes
    attr_reader :limit

    def initialize(limit)
      @flop = []
      @hands = []
      @hole_cards = []
      @holes = []
      @limit = limit
      @deck = PokerNode::Deck.new
    end

    def shuffle!
      @flop.fill(0, 3) { @deck.burn! }
      @turn_card = @deck.burn!
      @river_card = @deck.burn!
      @limit.times { @hole_cards << deal_hole }
      @holes = @hole_cards.collect { |hole| build_hole(hole) }
      @hands = @holes.map(&:hand)
    end

    def winning_hand
      @hands.max
    end

    def winning_hand_index
      @hands.index(winning_hand)
    end

    def inspect
      "<Deal\n\tflop=#{flop}\n\tturn_card=#{turn_card}\n\triver_card=#{river_card}>"
    end

    def deal_hole
      [].fill(0, 2) { @deck.burn! }
    end

    def build_hole(hole)
      PokerNode::Holdem::Hole.new(river, hole)
    end
  end
end

File: js/pokernode/deal.js
Deal = function(size) {
    deck = new Deck()
    this.flop = []
    for (var i = 0; i < 3; i++) {
        this.flop.push(deck.burn())
    }
    this.turn_card = deck.burn()
    this.turn = function() {
        return this.flop.concat([this.turn_card])
    }
    this.river_card = deck.burn()
    this.river = function() {
        return this.turn().concat([this.river_card])
    }
    this.size = size
    this.hole_cards = []
    for (i = 0; i < this.size; i++) {
        var cards = [deck.burn(), deck.burn()]
        this.hole_cards.push(cards)
    }
    this.hands = []
    for (i = 0; i < this.size; i++) {
        var hole = this.hole_cards[i]
        var hand = new Hand(this.river().concat(hole))
        this.hands.push(hand)
    }
    
    this.toString = function() {
        var holes = ""
        for (var i = 0; i < this.size; i++) {
            var hole = this.hole_cards[i]
            Hand.detect(this.hands[i])
            holes += "\thole " + (i + 1) + ": " + hole.toString() +
                " | " + this.hands[i].value + "\n"
        }
        return "river: " + this.river().toString() + "\n" + holes
    }
}
